internal/controller/instaslice_controller.go

```diff
@@ -91,19 +93,29 @@ func (r *InstasliceReconciler) Reconcile(ctx context.Context, req ctrl.Request)
 
 	isPodGated := checkIfPodGatedByInstaSlice(pod)
 
-	if !isPodGated && !controllerutil.ContainsFinalizer(pod, "org.instaslice/accelarator") {
+	if !isPodGated && !controllerutil.ContainsFinalizer(pod, finalizerOrGateName) {
 		//log.FromContext(ctx).Info("Ignoring ", "pod", pod.Name)
 		return ctrl.Result{}, nil
 	}
 
+	// Add finalizer to the pod gated by InstaSlice
+	if isPodGated && !controllerutil.ContainsFinalizer(pod, finalizerOrGateName) {
+		pod.Finalizers = append(pod.Finalizers, finalizerOrGateName)
+		errAddingFinalizer := r.Update(ctx, pod)
+		if errAddingFinalizer != nil {
+			log.FromContext(ctx).Error(errAddingFinalizer, "failed to add finalizer to pod")
+			return ctrl.Result{Requeue: true}, nil
+		}
+	}
+
 	var instasliceList inferencev1alpha1.InstasliceList
 
 	if err := r.List(ctx, &instasliceList, &client.ListOptions{}); err != nil {
 		log.FromContext(ctx).Error(err, "Error listing Instaslice")
 	}
 
 	// pod is completed move allocation to deleting state and return
-	if pod.Status.Phase == v1.PodSucceeded && controllerutil.ContainsFinalizer(pod, "org.instaslice/accelarator") {
+	if pod.Status.Phase == v1.PodSucceeded && controllerutil.ContainsFinalizer(pod, finalizerOrGateName) {
```

Comment:
What about Failed?

Reply:
FYI, the code changes here are syntactic sugar. Failed pods have to be deleted by the user for the current design

Reply:
They can't delete them if the finalizer isn't removed though.

Reply:
Ah, you are correct, can you please open GitHub issue? users would have to do a finalizer null which is painful.

Reply:
FYI, I created here: https://github.com/openshift/instaslice-operator/issues/94